Fix/result query

MoonSeongKwang opened the pull request:
### 주요 변경 내용
1. findUserScopeFromeRDB, findContractScopeFromRDB의 chain에서의
wrapper를 삭제하였습니다. chain.cpp 외부에서 사용하지 않는 함수이므로
굳이 유지할 필요가 없습니다.

2. setUserScopePid, setContractScopePid를 calculatePid로 통합하였습니다.
다만, tag/var_info가 존재할 때의 처리를 아직 덜 고려해서 이 함수는 추후
수정될 수 있습니다.

3. queryUserScope를 전반적으로 수정하였습니다.
생성할 때만 var_type이 주어지며, 그로 인하여 기존의 수정일 때는
var_type을 찾는 연산이 필요합니다.
이런 분기 사항들을 고려하지 않고 뼈대만 만들어두고 올려뒀는데, 설계상의
구현에 가깝게 수정하였습니다.

4. 위의 함수에 사용되는 getVarType과, 기타 다른 쿼리 처리에서도 사용되는
scope에서의 var_owner와 var_name의 unique함을 체크하는 함수를
구현하였습니다.
두 함수는 var_owner와 var_name을 가지고 var_type이 여러 개 존재하지
않는지를 체크하는 행동은 똑같이 수행하며, 단지 getVarType는 결과로써
true/false가 아니라 얻어낸 var_type 정보를 리턴하는 것이므로 두 함수를
통일해서 구현했습니다. 다만, 쓰이는 장소와 상황이 다르므로 함수명은 따로
구현하여 처리할 때 혼동이 없도록 하였습니다.

TODO : queryUserScope 외에도 Contract와 특히 transfer를 고쳐야 합니다.

MoonSeongKwang committed:
# FIX - Contract ledger, etc.
contract ledger부분에서 정확하게 ledger로 저장이 안 되는 부분을
수정하였습니다.
user cert에서 unique값은 uid가 아니고 sn이므로 수정했습니다.
후추 user cert는 한 uid가 여러 시점의 인증서를 가지고 있을 수 있는 것을
반영해야 합니다.
그 외 사소한 수정이 있었습니다.

MoonSeongKwang committed:
# EDIT - pid process change WIP

MoonSeongKwang committed:
# EDIT - queryUserScope, etc.
 ### 변경 내용
1. findUserScopeFromeRDB, findContractScopeFromRDB의 chain에서의
wrapper를 삭제하였습니다. chain.cpp 외부에서 사용하지 않는 함수이므로
굳이 유지할 필요가 없습니다.

2. setUserScopePid, setContractScopePid를 calculatePid로 통합하였습니다.
다만, tag/var_info가 존재할 때의 처리를 아직 덜 고려해서 이 함수는 추후
수정될 수 있습니다.

3. queryUserScope를 전반적으로 수정하였습니다.
생성할 때만 var_type이 주어지며, 그로 인하여 기존의 수정일 때는
var_type을 찾는 연산이 필요합니다.
이런 분기 사항들을 고려하지 않고 뼈대만 만들어두고 올려뒀는데, 설계상의
구현에 가깝게 수정하였습니다.

4. 위의 함수에 사용되는 getVarType과, 기타 다른 쿼리 처리에서도 사용되는
scope에서의 var_owner와 var_name의 unique함을 체크하는 함수를
구현하였습니다.
두 함수는 var_owner와 var_name을 가지고 var_type이 여러 개 존재하지
않는지를 체크하는 행동은 똑같이 수행하며, 단지 getVarType는 결과로써
true/false가 아니라 얻어낸 var_type 정보를 리턴하는 것이므로 두 함수를
통일해서 구현했습니다. 다만, 쓰이는 장소와 상황이 다르므로 함수명은 따로
구현하여 처리할 때 혼동이 없도록 하였습니다.

MoonSeongKwang committed:
# EDIT - Add tag-related logic
tag가 존재할 경우 인자가 하나 더 많은 calculatePid를 부르도록 함수를
수정하였습니다.

blackironj commented on src/plugins/chain_plugin/chain.cpp:
```diff
@@ -732,35 +725,177 @@ bool Chain::queryRunContract(UnresolvedBlock &UR_block, nlohmann::json &option,
 
-    rdb_controller->checkUnique(key);
+    if (!checkUniqueVarName(var_owner, var_name, height, vec_idx)) {
+      logger::ERROR("Error in `calculatePid`, there was several same (var owner, var name).");
```
unique var Name 이 아니라면 여기서 empty string 을 return 해야할거같은데 맞나요?
아니면 아래에서 계산된 pid가 return 될거 같습니다.

MoonSeongKwang replied:
아 그러네요. 수정하겠습니다.

blackironj commented on src/plugins/chain_plugin/chain.cpp:
```diff
@@ -732,35 +725,177 @@ bool Chain::queryRunContract(UnresolvedBlock &UR_block, nlohmann::json &option,
+
+bool Chain::checkUniqueVarName(const string &var_owner, const string &var_name, const block_height_type height, const int vec_idx) {
+  // 위 함수와 같은 동작을 하는데, 함수 명에서 혼동이 있을 수 있을 것 같아서 별개로 선언
+  return getVarType(var_owner, var_name, height, vec_idx);
```
`getVarType()` 이  int를 return  하는데 해당 함수는 boolean을 return하네요 혹시 의도하신 부분인가요?

MoonSeongKwang replied:
함수 여러차례 구현내용 바꾸다가 빠뜨린 모양이네요.
음.. 그런데 함수의 역할을 생각하면 bool이 적합할수도 있겠네요. 대신 지금처럼 그대로 return이 아니고 if문 넣어서 값 체크를 하긴 해야겠지만요. 좀 생각해보겠습니다.

MoonSeongKwang committed:
# FIX - calculatePid, etc.
1. tag/var_info가 주어지지 않을 때의 calculatePid에서 checkUnique한 결과가
false일 때, empty string을 return하도록 수정하였습니다.

2. tag/var_info까지 주어지는 경우의 calculatePid의 경우, checkUnique가
불필요합니다. pid를 계산할 모든 요소가 있기 때문입니다. 그래서
checkUnique를 하는 부분을 삭제하였고, 따라서 함수 인자에 checkUnique에서
쓰이기 위해 필요했던 height와 vec_idx가 삭제되었습니다.

3. checkUnique를 boolen 형태 return하는 함수로 확실히 구현하였습니다.
함수의 명을 별개로 유지하기로 한 이상, 동작까지 명확한 편이 좋을
것이므로 if문으로 bool return 하도록 하였습니다.

morishjs commented on src/plugins/chain_plugin/rdb_controller.cpp:
```diff
@@ -397,20 +397,46 @@ string RdbController::getUserCert(const base58_type &user_id) {
+  }
+}
+
+int RdbController::checkUniqueVarNameFromRDB(const string &var_owner, const string &var_name) {
```
`checkUniqueVarNameFromRDB` 함수 필요없어 보입니다. 사용하는 곳도 없고, 굳이 같은 기능을 하는 함수가 2개 존재할 필요가 없는 것 같아요.

MoonSeongKwang replied:
그러네요. 수정하겠습니다.

SeongKwangMoon commented:
@blackironj @morishjs 
getVarType / checkUniqueVarName 에 대해서 설명을 좀 써둬야 할 것 같아 추가합니다.

다른 종류의 임시값, ledger들은 특정 블록을 기준으로 최신의 정보만 있으면 충분합니다.
그러므로, head가 제대로 위치해있다면 state tree를 pid로 검색하기만 하면 바로 알 수 있습니다.
head를 옮기기 위해 현재 블록에 있는 값을 취소하고 이전으로 되돌아갈 때 또한 state tree의 값은 사용하지 못하지만(이 연산 자체가 state tree를 바꾸기 위한 사전 작업이기 때문입니다), 역시 특정 pid의 최신 값만 있으면 됩니다.

하지만 getVarType / checkUniqueVarName 는 다릅니다.
두 함수 모두 var_type이라는 요소가 부족한 상태에서 실행되는 함수이며, var_type는 pid를 계산하기 위해 필요한 요소입니다. 즉, pid를 알 수 없습니다. 그러므로 state tree에서 직접적인 검색도 불가능합니다.

var_owner와 var_name 두 가지가 일치하는 것이 유일하거나 없어야 한다는 조건입니다.
state tree에는 현재 head가 가리키고 있는 모든 정보가 들어있으므로, 분명 state tree에서 찾으면 원하는 정보가 존재하기는 합니다.
하지만 state tree는 pid가 없으면 모든 트리를 순회해야하는 구조인데, 하나하나 접근하여 값을 찾기에는 너무 비효율적입니다.

그래서 이 PR에서의 getVarType / checkUniqueVarName 는 현 블록에서부터 RDB에 반영되기 전의 블록까지 거슬러올라가며 모든 ledger를 for로 begin부터 end까지 손수 찾는 것이며, RDB에서 query를 통해 빠르게 원하는 정보를 검색하는 방식으로 구현되었습니다.

함수 구현에 대한 설명을 남겨둘 겸, 함수를 체크하실 때 조금이라도 도움이 되시라고 씁니다.

morishjs commented on src/plugins/chain_plugin/rdb_controller.cpp:
```diff
@@ -397,20 +397,46 @@ string RdbController::getUserCert(const base58_type &user_id) {
-  else
-    return false;
+    // clang-format off
+    if(var_owner.size() > 45)
```
`var_owner.size() > 45` 보다는 함수로 하나 만들어서 (예, isUserScope) 의도를 분명하게 하는게 좋을 것 같습니다.

morishjs replied:
그리고 user_scope와 contract_scope 를 구분하는 기준이 var_owner의 길이를 기준으로 하는건가요?

MoonSeongKwang replied:
우선은 러프하게 구현하느라 이렇게 해뒀는데, 다음번 수정할 때 일괄적으로 고치겠습니다.
var_owner의 길이를 기준으로만 구별하면 되는지에 대해서는 아직 깊이 생각해보지 않았습니다.

그게 직관적으로는 가장 구분하기 편해서 그런건데, 뭔가 다른 validation이 필요하다고 판단되면 isUserScope 함수에 추후 구현해서 넣으면 될 것 같네요. TODO로 두겠습니다.
